### PR TITLE
feature/Game

### PR DESCRIPTION
**Pull request by peenyaa7**
## Description

[//]: # (Please inlcude a summary of the change and which issues is fixed. Please also include relavant motivation and context. List any dependencies that are required fo change.)

Add operations related to the game

## Trello link:

[Trello card](https://trello.com/c/eTnxNqI4)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist
[//]: # (Go over all the following points, and put an `x` in all the boxes that apply.)
[//]: # (If you're unsure about any of these, don't hesitate to ask.)

- [ ] Changes were successfully tested.
- [ ] Dependencies. All dependencies are deployed.

## How Has This Been Tested?

[//]: # (Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration)

Manually, using https://jiangxy.github.io/websocket-debug-tool/

### CODE REVIEW

File: src/main/java/com/devathon/preguntonic/config/WebSocketConfig.java [hunk 1]
```diff
@@ -37,9 +37,8 @@ public void registerStompEndpoints(StompEndpointRegistry registry) {
   @Override
   public void configureMessageBroker(MessageBrokerRegistry registry) {
     registry
-        .setApplicationDestinationPrefixes("/app") // [1]The prefix for the application destination.
-        .enableSimpleBroker(
-            "/room"); // [2]Enable the simple broker (some examples of destinations).
+        .setApplicationDestinationPrefixes("/preguntonic") // [1]The prefix for the application destination.
+        .enableSimpleBroker("/"); // [2]Enable the simple broker (some examples of destinations).
```
Comment (JFCowboy) [inline src/main/java/com/devathon/preguntonic/config/WebSocketConfig.java]: Why is this change in the broker suggested? 
if it is needed for more `topics` isn't it possible to do something like: 

```suggestion
        .setApplicationDestinationPrefixes("/preguntonic") // [1]The prefix for the application destination.
        .enableSimpleBroker("/room", "/chat", "/topic"); // [2]Enable the simple broker (some examples of destinations).
```

**Comment by JFCowboy** [comment]
Remember run:
`mvn spotless:apply`
Before open the PR

Also I added a new githook in this PR but i'm not sure it works in other OS
https://github.com/Devathon-2024-team5/Preguntonic-backend/pull/15

File: src/main/java/com/devathon/preguntonic/constants/Constants.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.devathon.preguntonic.constants;
+
+public class Constants {
```
Comment (JFCowboy) [inline src/main/java/com/devathon/preguntonic/constants/Constants.java]: I wonder if this class is really necessary?
I think they are very specific constants for a class, I would propose either moving the constants to the specific class or changing the name to something like GameConstants

File: src/main/java/com/devathon/preguntonic/config/WebSocketConfig.java (hunk 1 above)
Comment (JFCowboy) [reply src/main/java/com/devathon/preguntonic/config/WebSocketConfig.java]: anyway this also affect the other controller... but depending of which PR is merge first we can resolve conflicts

File: src/main/java/com/devathon/preguntonic/controllers/GameController.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    void gameSubscription(@DestinationVariable String code);
+
+    /**
+     * Players send when they join the game
+     * @param code Code of the room
+     * @return GameStatusDto with the updated players
+     */
+    @MessageMapping("/players/{playerId}/join")
+    @SendTo("/preguntonic/rooms/{code}/game")
+    public GameStatusDto joinGame(@DestinationVariable String code, @DestinationVariable UUID playerId);
```
Comment (JFCowboy) [inline src/main/java/com/devathon/preguntonic/controllers/GameController.java]: are we always going to use UUID?, to change this in the my PR for the controller.

File: src/main/java/com/devathon/preguntonic/controllers/GameController.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import org.springframework.messaging.handler.annotation.SendTo;
+import org.springframework.messaging.simp.annotation.SubscribeMapping;
+
+import com.devathon.preguntonic.dto.GameStatusDto;
+import com.devathon.preguntonic.dto.PlayerQuestionResponseDto;
+import com.devathon.preguntonic.dto.QuestionResultDto;
+
+
+@MessageMapping("/rooms/{code}/game")
+public interface GameController {
```
Comment (JFCowboy) [inline src/main/java/com/devathon/preguntonic/controllers/GameController.java:19]: 👍

File: src/main/java/com/devathon/preguntonic/model/Room.java [hunk 5]
```diff
@@ -23,7 +25,9 @@ public class Room {
   private RoomStatus status;
   private Game game;
   private int maxPlayers;
-  private int currentPlayers;
   private int numQuestions;
   private LocalDateTime createdAt;
+
+  private final Map<UUID, Player> players = Map.of();
```
Comment (JFCowboy) [inline src/main/java/com/devathon/preguntonic/model/Room.java:30]: This can cause conflicts, but I can use Map instead of List

File: src/main/java/com/devathon/preguntonic/model/PlayerStatus.java [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package com.devathon.preguntonic.model;
+
+public enum PlayerStatus {
```
Comment (JFCowboy) [inline src/main/java/com/devathon/preguntonic/model/PlayerStatus.java:8]: I like this PlayerStatus, but WDYT to also used it in `Lobby` to handle the `is_ready` state and also remove this boolean param.

File: src/main/java/com/devathon/preguntonic/model/Player.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+@Data
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class Player {
+    
+    private UUID id;
+    private boolean isAdmin;
+    private String ipAddress;
+    private String avatar; // TODO: Change to enum if needed
```
Comment (JFCowboy) [inline src/main/java/com/devathon/preguntonic/model/Player.java]: 👍 I like this idea, but we need to coordinate with frontend

File: src/main/java/com/devathon/preguntonic/model/Player.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    private int score;
+
+    // Transient (lobby)
+    private boolean isReady;
+
+    // Transient (game)
+    private boolean responded;
+    private long responseTime;
+    private UUID responseId;
+    private boolean readyForNextQuestion;
```
Comment (JFCowboy) [inline src/main/java/com/devathon/preguntonic/model/Player.java]: NIT: in a future we can extract this parameter to other class due it is more related with game logic

File: src/main/java/com/devathon/preguntonic/dto/PlayerStatusDto.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+import lombok.Builder;
+
+@Builder
+public class PlayerStatusDto {
+        
+    @JsonProperty("id")
+    private UUID id;
+
+    @JsonProperty("nickname")
+    private String nickname;
```
Comment (JFCowboy) [inline src/main/java/com/devathon/preguntonic/dto/PlayerStatusDto.java]: please confirm with front-end if we are really going to use the nickname attribute here or just name.
@NicolasLLans @Alejandro25AR @WanderleeDev

File: src/main/java/com/devathon/preguntonic/controllers/GameControllerImpl.java [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+import com.devathon.preguntonic.model.Room;
+import com.devathon.preguntonic.model.RoomStatus;
+import com.devathon.preguntonic.services.DummyRoomService;
+import com.devathon.preguntonic.services.QuestionService;
+
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+@Controller
+public class GameControllerImpl implements GameController {
```
Comment (JFCowboy) [inline src/main/java/com/devathon/preguntonic/controllers/GameControllerImpl.java:32]: NIT: if you have time we can handle a layered architecture where the controller calls a service layer.

Thus, the controller's only responsibility would be to manage the request and parameters that arrive and coordinate/orchestrate the calls to the services.

So for now the proposed architecture is:
Controller(handle api calls)-->Services(Room/Game logic)-->Repositories(database access)

WDYT?

File: src/main/java/com/devathon/preguntonic/controllers/GameControllerImpl.java [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+
+    @Autowired
+    private DummyRoomService roomService;
```
Comment (JFCowboy) [inline src/main/java/com/devathon/preguntonic/controllers/GameControllerImpl.java]: In this case could be better if we use the Interface instead of concrete implementation.
```suggestion
    private RoomService roomService;
```

File: src/main/java/com/devathon/preguntonic/controllers/GameControllerImpl.java [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+            return null;
+        }
+
+        Room room = optRoom.get();
+
+        // Check if the player is already in the room
+        if (!room.getPlayers().containsKey(playerId)) {
+            log.error("Player {} is not in the room {}", playerId, code);
+            return null;
+        }
```
Comment (JFCowboy) [inline src/main/java/com/devathon/preguntonic/controllers/GameControllerImpl.java]: I think we need in Room service a getPlayers Method that handle this validations I can add it in this PR https://github.com/Devathon-2024-team5/Preguntonic-backend/pull/16/

But do you prefer that the method return an Exception or null if Room or player not found? For now I prefer Exception but WDYT?

Comment (JFCowboy) [reply src/main/java/com/devathon/preguntonic/controllers/GameControllerImpl.java]: Also maybe you can use the current `Optional<BasicPlayer> getPlayer(String roomCode, int playerId)` maybe should i change the returned object?

File: src/main/java/com/devathon/preguntonic/config/WebSocketConfig.java (hunk 1 above)
Comment (peenyaa7) [reply src/main/java/com/devathon/preguntonic/config/WebSocketConfig.java]: We only need `/rooms` broker, so we can use that in all topics

File: src/main/java/com/devathon/preguntonic/constants/Constants.java (hunk 2 above)
Comment (peenyaa7) [reply src/main/java/com/devathon/preguntonic/constants/Constants.java]: If u are more confortable with that we can move constants to each controller 👌

File: src/main/java/com/devathon/preguntonic/controllers/GameController.java (hunk 3 above)
Comment (peenyaa7) [reply src/main/java/com/devathon/preguntonic/controllers/GameController.java]: I prefer use UUID, using that we can avoid possible conflicts

File: src/main/java/com/devathon/preguntonic/model/Room.java (hunk 5 above)
Comment (peenyaa7) [reply src/main/java/com/devathon/preguntonic/model/Room.java:30]: Why? We are not using something like that i this class

File: src/main/java/com/devathon/preguntonic/model/PlayerStatus.java (hunk 6 above)
Comment (peenyaa7) [reply src/main/java/com/devathon/preguntonic/model/PlayerStatus.java:8]: We can add `PlayerStatus.READY` to do that. WDYT

File: src/main/java/com/devathon/preguntonic/model/Player.java (hunk 8 above)
Comment (peenyaa7) [reply src/main/java/com/devathon/preguntonic/model/Player.java]: Yes... i think we should maintain those params until MVP is released

File: src/main/java/com/devathon/preguntonic/controllers/GameControllerImpl.java (hunk 10 above)
Comment (peenyaa7) [reply src/main/java/com/devathon/preguntonic/controllers/GameControllerImpl.java:32]: It is a fairly good idea, but, what logic will be in the controller and what in the service? (null checks, logic, etc.)

File: src/main/java/com/devathon/preguntonic/controllers/GameControllerImpl.java (hunk 12 above)
Comment (peenyaa7) [reply src/main/java/com/devathon/preguntonic/controllers/GameControllerImpl.java]: We have to coordinate to do that. We can use room service methods as well. About exception... it is a good idea, but, how do we could return this exceptions to the frontend?? For this reason i prefer an object that is always returned and filled with the good/bad results

File: src/main/java/com/devathon/preguntonic/controllers/GameController.java (hunk 3 above)
Comment (JFCowboy) [reply src/main/java/com/devathon/preguntonic/controllers/GameController.java]: What I don't like about this is how long the url would be (especially for testing) but I wouldn't have a problem changing it in the other controllers.

@Alejandro25AR @NicolasLLans @WanderleeDev WDYT?

File: src/main/java/com/devathon/preguntonic/model/Room.java (hunk 5 above)
Comment (JFCowboy) [reply src/main/java/com/devathon/preguntonic/model/Room.java:30]: I was referring to the change in this [PR](https://github.com/Devathon-2024-team5/Preguntonic-backend/pull/9/files) in that I start using only a list, but I can change it in the other PR

File: src/main/java/com/devathon/preguntonic/model/PlayerStatus.java (hunk 6 above)
Comment (JFCowboy) [reply src/main/java/com/devathon/preguntonic/model/PlayerStatus.java:8]: I like it 👍

File: src/main/java/com/devathon/preguntonic/controllers/GameControllerImpl.java (hunk 10 above)
Comment (JFCowboy) [reply src/main/java/com/devathon/preguntonic/controllers/GameControllerImpl.java:32]: Mainly this:

Controller(handle api calls and responses)
- orchestrate Services calls
- build/transform service responses to API expected...

Services(Room/Game logic)
- handle the logic of the game

File: src/main/java/com/devathon/preguntonic/controllers/GameController.java (hunk 3 above)
Comment (NicolasLLans) [reply src/main/java/com/devathon/preguntonic/controllers/GameController.java]: If you have to add the Player_ID to the request, I think there will be no problems on the frontend side

File: src/main/java/com/devathon/preguntonic/dto/PlayerStatusDto.java (hunk 9 above)
Comment (NicolasLLans) [reply src/main/java/com/devathon/preguntonic/dto/PlayerStatusDto.java]: Only playername.

File: src/main/java/com/devathon/preguntonic/controllers/GameController.java (hunk 3 above)
Comment (JFCowboy) [inline src/main/java/com/devathon/preguntonic/controllers/GameController.java]: After test it in docker and connecting with front end, we should add the name of the "Path" variable
```suggestion
    public GameStatusDto joinGame(@DestinationVariable("code") String code, @DestinationVariable("playerId") UUID playerId);
```

File: src/main/java/com/devathon/preguntonic/controllers/GameController.java [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    public GameStatusDto joinGame(@DestinationVariable String code, @DestinationVariable UUID playerId);
+
+    /**
+     * Players send when they answer a question
+     * @param code Code of the room
+     * @return QuestionResultDto if all players answered, null otherwise
+     */
+    @MessageMapping("/players/{playerId}/response")
+    @SendTo("/preguntonic/rooms/{code}/game")
+    public QuestionResultDto answerQuestion(@DestinationVariable String code, @DestinationVariable UUID playerId, @Payload PlayerQuestionResponseDto response);
```
Comment (JFCowboy) [inline src/main/java/com/devathon/preguntonic/controllers/GameController.java]: Same here
```suggestion
    @MessageMapping("/players/{playerId}/response")
    @SendTo("/preguntonic/rooms/{code}/game")
    public QuestionResultDto answerQuestion(@DestinationVariable("code") String code, @DestinationVariable("playerId") UUID playerId, @Payload PlayerQuestionResponseDto response);
```

File: src/main/java/com/devathon/preguntonic/controllers/GameController.java (hunk 3 above)
Comment (peenyaa7) [reply src/main/java/com/devathon/preguntonic/controllers/GameController.java]: The problem is the format. UUID is too long, but is unique, so u need to install something to manage this data type on frontend.

If there is no problem, resolve that 👌

File: src/main/java/com/devathon/preguntonic/model/Room.java (hunk 5 above)
Comment (peenyaa7) [reply src/main/java/com/devathon/preguntonic/model/Room.java:30]: Thats true... i implement it on this way to execute the life cycle of the game. We can use anyone, but i prefer Map because is O(1) on "get" operations

File: src/main/java/com/devathon/preguntonic/constants/Constants.java (hunk 2 above)
Comment (peenyaa7) [reply src/main/java/com/devathon/preguntonic/constants/Constants.java]: Already moved 👍

File: src/main/java/com/devathon/preguntonic/config/WebSocketConfig.java [hunk 14]
```diff
@@ -37,9 +37,10 @@ public void registerStompEndpoints(StompEndpointRegistry registry) {
   @Override
   public void configureMessageBroker(MessageBrokerRegistry registry) {
     registry
-        .setApplicationDestinationPrefixes("/app") // [1]The prefix for the application destination.
+        .setApplicationDestinationPrefixes(
+            "/preguntonic") // [1]The prefix for the application destination.
         .enableSimpleBroker(
-            "/room"); // [2]Enable the simple broker (some examples of destinations).
+            "/rooms"); // [2]Enable the simple broker (some examples of destinations).
```
Comment (JFCowboy) [inline src/main/java/com/devathon/preguntonic/config/WebSocketConfig.java]: we are doing the frontend integration with previous version, So I propose make this change in another PR, so as not to delay the integration further and do this cleaning when we have more functionalities running.

Comment (peenyaa7) [reply src/main/java/com/devathon/preguntonic/config/WebSocketConfig.java]: But in yours version u are using `/rooms`, so i change to `/rooms`
image

**Review by JFCowboy — APPROVED** [review]
